// @telekom/scale-components/dist/cjs/scale-notification-banner.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-53b94a75.js');
const index$1 = require('./index-53f5a5fc.js');
const statusNote = require('./status-note-b89242d6.js');

const notificationBannerCss = ":host{--width:100%;--radius:var(--scl-radius-8);--background-error:hsla(353.79999999999995, 100%, 88.6%, 0.35);--background-warning:hsla(33.5, 100%, 75.1%, 0.3);--background-informational:hsla(190.20000000000005, 100%, 92%, 0.75);--background-success:hsl(98.80000000000001, 49.7%, 61%, 0.2)}.notification-banner{border-radius:var(--radius);width:var(--width);position:relative;box-shadow:var(--scl-shadow-level-1);min-height:48px}.notification-banner--variant-error{background-color:var(--background-error)}.notification-banner--variant-warning{background-color:var(--background-warning)}.notification-banner--variant-informational{background-color:var(--background-informational)}.notification-banner--variant-success{background-color:var(--background-success)}.notification-banner__heading{margin:14px 48px 14px 48px;font-weight:var(--scl-font-weight-bold);line-height:48px}.notification-banner__text ::slotted(*){margin:-10px 0px -25px 0px;line-height:16px;padding-bottom:15px}.notification-banner--has-no-text .notification-banner__link ::slotted(*){margin:0;padding:0px 0px 0px 16px}.notification-banner--has-text .notification-banner__link ::slotted(*){margin:20px 0 0 0;padding-bottom:15px;line-height:20px}::slotted(*){font-weight:var(--scl-font-weight-regular)}.notification-banner__button-close{position:absolute;top:8px;right:13.5px;color:#191919;border:none;cursor:pointer;margin:0;padding:0;background:transparent}.notification-banner__button-close svg{height:19px;width:19px;padding:6.5px;border-radius:20%}.notification-banner__button-close:hover svg{background-color:white;color:var(--scl-color-primary-hover)}.notification-banner__icon-success{position:absolute;top:12.5px;left:17px;height:20px;width:20px;color:var(--scl-color-background-success)}.notification-banner__icon-error{position:absolute;top:12px;left:17px;height:20px;width:20px;color:var(--scl-color-background-error)}.notification-banner__icon-information{position:absolute;top:12px;left:17px;height:20px;width:20px;color:var(--scl-color-text-info)}@media screen and (forced-colors: active), (-ms-high-contrast: active){.notification-banner__icon-close{color:white}}";

const NotificationBanner = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.variant = 'informational';
    this.dismissible = false;
    this.autoHide = false;
    this.autoHideDuration = 3000;
    this.close = () => {
      this.opened = false;
    };
  }
  componentWillLoad() {
    this.hasSlotText = !!this.hostElement.querySelector('[slot=text]');
    this.hasSlotLink = !!this.hostElement.querySelector('[slot=link]');
  }
  componentDidUpdate() {
    this.hasSlotText = !!this.hostElement.querySelector('[slot=text]');
    this.hasSlotLink = !!this.hostElement.querySelector('[slot=link]');
  }
  connectedCallback() {
    statusNote.statusNote({ source: this.hostElement, type: 'warn' });
    if (this.autoHide === true) {
      setTimeout(this.close, this.autoHideDuration);
    }
  }
  async open() {
    this.opened = true;
  }
  handleIcons() {
    if (this.variant) {
      switch (this.variant) {
        case 'success':
          return (index.h("scale-icon-alert-success", { class: "notification-banner__icon-success", color: "#187431", "accessibility-title": "success" }));
        case 'informational':
          return (index.h("scale-icon-alert-information", { class: "notification-banner__icon-information", "accessibility-title": "information" }));
        case 'error':
          return (index.h("scale-icon-alert-error", { class: "notification-banner__icon-error", "accessibility-title": "error" }));
        case 'warning':
          return (index.h("scale-icon-alert-error", { class: "notification-banner__icon-information", color: "#AE461C", "accessibility-title": "information" }));
      }
    }
    return;
  }
  render() {
    if (!this.opened) {
      return null;
    }
    return (index.h(index.Host, null, index.h("div", { part: this.getBasePartMap(), class: this.getCssClassMap(), tabindex: "0" }, index.h("div", { part: "container", class: "notification-banner__container" }, this.handleIcons(), index.h("div", { part: "heading", class: "notification-banner__heading" }, index.h("slot", null), this.dismissible && (index.h("button", { part: "button-dismissable", class: "notification-banner__button-close", onClick: () => this.close(), onKeyDown: (e) => {
        if (e.key === 'Enter') {
          this.close();
        }
      } }, index.h("scale-icon-action-circle-close", { "accessibility-title": "close" }))), this.hasSlotText && (index.h("div", { part: "text", class: "notification-banner__text" }, index.h("slot", { name: "text" }))), this.hasSlotLink && (index.h("scale-link", { href: this.href, class: "notification-banner__link" }, index.h("slot", { name: "link" }))))))));
  }
  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }
  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }
  getCssOrBasePartMap(mode) {
    const name = 'notification-banner';
    const prefix = mode === 'basePart' ? '' : `${name}--`;
    return index$1.classnames(name, this.variant && `${prefix}variant-${this.variant}`, this.hasSlotText && `${prefix}has-text`, !this.hasSlotText && `${prefix}has-no-text`);
  }
  get hostElement() { return index.getElement(this); }
};
NotificationBanner.style = notificationBannerCss;

exports.scale_notification_banner = NotificationBanner;
