// @telekom/scale-components/dist/cjs/scale-card.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-53b94a75.js');
const index$1 = require('./index-53f5a5fc.js');

const cardCss = ":host{--background:var(--scl-color-background-standard);--transition:all var(--scl-motion-duration-fast)\n    var(--scl-motion-easing-standard);--radius:var(--scl-radius-12);--box-shadow:var(--scl-shadow-level-1);--box-shadow-hover:var(--scl-shadow-level-1-hover);--box-shadow-focus:0 0 0 var(--scl-spacing-2) var(--scl-color-focus);--box-shadow-active:var(--scl-shadow-level-1-active);--spacing-body:var(--scl-spacing-24);--spacing-body-slotted:0;--spacing-body-slotted-interactive:0}.card-border{border:1px solid transparent}.card{width:100%;overflow:hidden;box-sizing:border-box;background:var(--background);transition:var(--transition);border-radius:var(--radius);box-shadow:var(--box-shadow)}.card__body{padding:var(--spacing-body)}.card__body ::slotted(*){margin:var(--spacing-body-slotted)}.card--interactive{color:inherit;cursor:pointer;display:block;outline:none;text-decoration:none}.card--interactive:hover{box-shadow:var(--box-shadow-hover)}.card--interactive:focus{box-shadow:var(--scl-shadow-level-1-hover), var(--box-shadow-focus)}.card--interactive:active{border:none;box-shadow:var(--box-shadow-active)}.card--interactive .card__body ::slotted(*){margin:var(--spacing-body-slotted-interactive)}";

const Card = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /** (optional) Link card */
    this.to = '';
    /** (optional) Label of the card */
    this.label = '';
    /** (optional) Link card target */
    this.target = '_self';
    /** (optional) Link card rel */
    this.rel = '';
  }
  render() {
    const Tag = !!this.to ? 'a' : 'div';
    return (index.h(index.Host, null, this.styles && index.h("style", null, this.styles), index.h("div", { class: "card-border", part: "border" }, index.h(Tag, Object.assign({ class: this.getCssClassMap(), part: index$1.classnames('base', !!this.to && 'interactive'), onClick: this.clickCard }, (!this.to ? { role: 'group' } : {}), (!!this.to ? { href: this.to } : {}), (!!this.target ? { target: this.target } : {}), (!!this.rel ? { rel: this.rel } : {}), (!!this.label ? { ['aria-label']: this.label } : {})), index.h("div", { class: "card__body", part: "body" }, index.h("slot", null))))));
  }
  getCssClassMap() {
    return index$1.classnames('card', !!this.to && 'card--interactive');
  }
};
Card.style = cardCss;

exports.scale_card = Card;
