// @telekom/scale-components/dist/cjs/scale-list-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-53b94a75.js');
const index$1 = require('./index-53f5a5fc.js');

const listItemCss = ":host(scale-list-item){--display:block;--font-size:var(--scl-font-size-16);--line-height:var(--scl-font-line-height-150);--spacing-left:calc(0.25ch + var(--scl-spacing-16));--font-size-marker-ordered:var(--scl-font-size-16);--font-weight-marker-ordered:var(--scl-font-weight-bold);--font-size-marker-ordered-nested:var(--scl-font-size-12);--line-height-marker-ordered-nested:var(--scl-font-line-height-150);--spacing-top-marker-ordered-nested:calc(\n    var(--scl-spacing-16) - var(--scl-spacing-12)\n  );--spacing-right-no-marker:var(--scl-spacing-8);display:var(--display)}.list-item{position:relative;font-size:var(--font-size);line-height:var(--line-height);padding-left:var(--spacing-left)}.list-item:before{content:'';position:absolute;top:0;left:0}.list-item--unordered:before{top:0.5em;border:var(--scl-spacing-1) solid currentColor;display:block;transform:scale(0.66);background:currentColor;box-sizing:border-box;border-radius:50%;width:var(--scl-spacing-8);height:var(--scl-spacing-8)}.list-item--nested.list-item--unordered:before{border:var(--scl-spacing-1) solid currentColor;background:transparent}.list-item--ordered:before{content:attr(data-index) '.';font-size:var(--font-size-marker-ordered);font-weight:var(--font-weight-marker-ordered)}.list-item--nested.list-item--ordered:before{font-size:var(--font-size-marker-ordered-nested);line-height:var(--line-height-marker-ordered-nested);padding-top:var(--spacing-top-marker-ordered-nested)}.list-item--no-marker{display:inline-flex;align-items:center;padding-left:0}.list-item--no-marker:before{display:none}.list-item--no-marker>::slotted(*){margin-right:var(--spacing-right-no-marker)}";

const ListItem = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /** Whether this is a child of an ordered scale-list, gets set automatically by its parent */
    this.ordered = false;
    /** If `false`, no marker or left padding will be visible */
    this.marker = true;
    this.hasNestedChild = false;
    this.isNested = false;
    this.handleSlotChange = ({ target }) => {
      this.hasNestedChild =
        target.assignedNodes().length > 0;
      this.isNested = this.isNestedCheck();
    };
    this.isNestedCheck = () => {
      return this.el.closest('scale-list[slot="nested"]') != null;
    };
  }
  componentWillLoad() {
    this.isNested = this.isNestedCheck();
  }
  connectedCallback() {
    if (!this.el.hasAttribute('role')) {
      this.el.setAttribute('role', 'listitem');
    }
  }
  render() {
    return (index.h(index.Host, null, this.styles && index.h("style", null, this.styles), index.h("div", { class: this.getCssClassMap(), "data-index": this.index, part: index$1.classnames('base', this.ordered ? 'ordered' : 'unordered', this.isNested && 'nested', !this.marker && 'no-marker') }, index.h("slot", null), index.h("div", { class: "list-item__nested-list", part: "nested-list", hidden: !this.hasNestedChild }, index.h("slot", { name: "nested", onSlotchange: this.handleSlotChange })))));
  }
  getCssClassMap() {
    const orderType = this.ordered ? 'ordered' : 'unordered';
    return index$1.classnames('list-item', this.isNested && 'list-item--nested', `list-item--${orderType}`, !this.marker && 'list-item--no-marker');
  }
  get el() { return index.getElement(this); }
};
ListItem.style = listItemCss;

exports.scale_list_item = ListItem;
