// @telekom/scale-components/dist/esm/scale-logo.entry.js
import { r as registerInstance, h, a as Host, g as getElement } from './index-f2e5998d.js';
import { c as classnames } from './index-713f92a5.js';

const logoCss = ":host{--logo-size:36px}.logo{display:inline-flex;height:var(--logo-size);padding:calc(var(--logo-size) / 3);position:relative}.logo--variant-magenta{background-color:var(--scl-color-white)}.logo--variant-white{background-color:var(--scl-color-primary)}.logo--transparent{background-color:transparent}";

const Logo = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /** (optional) Variant/color of the logo text and logo */
    this.variant = 'magenta';
    /** (optional) Set transparent background */
    this.transparent = false;
    /** (optional) Language of the logo text/ claimOff showes just the T Logo */
    this.language = 'en';
    /** (optional) The height in pixels */
    this.size = 36;
    /** (optional) Set a link */
    this.href = '';
  }
  styles() {
    return `:host {
        --logo-size: ${this.size}px;
      }`;
  }
  getLogoSvg(role) {
    if (this.accessibilityTitle) {
      return (h("scale-logo-svg", { language: this.language, color: this.variant, size: this.size, accessibilityTitle: this.accessibilityTitle, role: role }));
    }
    else {
      return (h("scale-logo-svg", { language: this.language, color: this.variant, size: this.size, role: role }));
    }
  }
  render() {
    return (h(Host, null, h("style", null, this.styles()), this.href === '' ? (h("div", { class: this.getCssClassMap(), onClick: this.clickHandler }, this.getLogoSvg('img'))) : (h("a", { href: this.href, class: this.getCssClassMap(), onClick: this.clickHandler }, this.getLogoSvg('link')))));
  }
  getCssClassMap() {
    return classnames(`logo`, this.variant && `logo--variant-${this.variant}`, this.transparent && `logo--transparent`);
  }
  get hostElement() { return getElement(this); }
};
Logo.style = logoCss;

export { Logo as scale_logo };
